SPICE-0025: `pkl run` CLI framework (#1367)
apple/pkl: docs/modules/pkl-cli/pages/index.adoc

IMPORTANT: Users of `pkl run` must be aware of the security implications of this behavior.
Command line tools implemented in Pkl will be able to write or overwrite any file that the user has permissions to modify.
This matches the behavior of command line tools written in any other language.

Beyond writing to files (via `output.files`) and the standard output (via `output.text` or `output.bytes`), commands may not trigger side effects not already available in Pkl.
Pkl provides no facility for making arbitrary HTTP requests (i.e. non-`GET` methods or setting headers), executing other programs, or any other side effect.
If additional functionality is desired, xref:language-reference:index.adoc#external-readers[external readers] may be used to extends Pkl's capabilities.

=== Command Options

Each property of the class of a command's `options` property becomes a command line option.
Properties with the `local`, `hidden`, `fixed`, and/or `const` modifiers are not parsed as options
A property's doc comment, if present, becomes the corresponding option's CLI help description.
Doc comments are interpreted as Markdown text and formatted nicely when displayed to users.

> Properties must have xref:language-reference:index.adoc#type-annotations[type annotations] to determine how they are parsed.

Reviewer: What if they are not? I think it's worth clarifying.

By default, module outputs are separated with `---`, as in a YAML stream.
The separator can be customized using the `--module-output-separator` option.

[[cli-tools]]
== Implementing CLI Tools

CLI tools can be implemented in Pkl by modules extending the `pkl:Command` module.
These tools are executed by running `pkl run`.

Pkl commands have a few properties that distinguish them from standard module evaluation:

* Users provide input to commands using familiar command line idioms, providing a better experience than deriving inputs from xref:language-reference:index.adoc#resources[resources] like external properties or environment variables.
* Commands can produce output to standard output (via `output.text` or `output.bytes`) and written files (via `output.files`) at the same time.

> * Like CLI tools written in other languages, command file output may write to any absolute path (not just relative to the `--multiple-file-output-path` option).

Reviewer 2: Seems needlessly defensive phrasing. 

```suggestion
* Command file output may write to any absolute path (not only relative to the `--multiple-file-output-path` option).
```

* Commands provide a way to dynamically import modules when they are specified as command line options.

Commands are implemented as regular modules and declare their supported command line flags and positional arguments using a class with annotated properties.

=== Defining Commands

Commands are defined by creating a module that extends `pkl:Command`:

[source,pkl%tested]
.my-tool.pkl
----
/// This doc comment becomes part of the command's CLI help!
/// Markdown formatting is **allowed!**
extends "pkl:Command"

options: Options

class Options {
  // Define CLI flags/arguments...
}

// Regular module code...
----

Like `pkl eval`, when a command completes without an evaluation error the process exits successfully (exit code 0).

> Commands can return a failure using `throw` (exit code 1), but otherwise may not control the exit code.

Author: I considered this, but to do that we'd need a new `CommandOutput` class that subclassed `ModuleOutput` and I'm hesitant to `open` that up without being able to make `FileOutput` a `sealed` class.

This is a possible avenue for future enhancements if there is sufficient demand.

Commands may set `output.text`/`output.bytes` and/or `output.files`.
Unlike `pkl eval`, both outputs (stdout and files) are processed for all commands, so commands may write files and generate output.
Additionally, keys in `output.files` are not "sandboxed" by `--multiple-file-output-path`.
Instead, file entries with absolute paths are written to the specified path and entries with relative path are written relative to the current working directory (or `--working-dir` if specified).

> Written file paths are printed to the command's standard error.

Reviewer 2: A lot of the meaning of this paragraph overlaps with the bullets on how commands evaluate differently than modules, but worded slightly differently. This makes me feel I need to understand the differences in wording... and then I almost miss the last sentence, which is entirely new.

IMPORTANT: Users of `pkl run` must be aware of the security implications of this behavior.
Command line tools implemented in Pkl will be able to write or overwrite any file that the user has permissions to modify.

> This matches the behavior of command line tools written in any other language.

Reviewer 2: Overly defensive, especially the "others do it, too".... If you want to warn explicitly, I would not defend it by comparing to other languages, but by contrasting it with usual Pkl behaviour...

```suggestion
IMPORTANT: Users of `pkl run` must be aware of the security implications of this behavior. Whereas usual `pkl eval` prevents accidental overwrites, by not allowing absolute paths, `pkl run` does not offer this protection. As such, it writes the credentials of the invoking user.
```
